helix-core/src/main/java/org/apache/helix/controller/rebalancer/MaintenanceRebalancer.java

```diff
@@ -55,16 +59,46 @@ public IdealState computeNewIdealState(String resourceName, IdealState currentId
 
     // One principal is to prohibit DROP -> OFFLINE and OFFLINE -> DROP state transitions.
     // Derived preference list from current state with state priority
+    StateModelDefinition stateModelDef = clusterData.getStateModelDef(currentIdealState.getStateModelDefRef());
+
     for (Partition partition : currentStateMap.keySet()) {
       Map<String, String> stateMap = currentStateMap.get(partition);
       List<String> preferenceList = new ArrayList<>(stateMap.keySet());
+
+      // This sorting preserves the ordering of current state hosts in the order of current IS pref list
       Collections.sort(preferenceList, new PreferenceListNodeComparator(stateMap,
-          clusterData.getStateModelDef(currentIdealState.getStateModelDefRef()),
-          Collections.<String>emptyList()));
+          stateModelDef, currentIdealState.getPreferenceList(partition.getPartitionName())));
+
+      // This sorts the current state hosts based on the priority
+      preferenceList.sort(new StatePriorityComparator(stateMap, stateModelDef));
+
       currentIdealState.setPreferenceList(partition.getPartitionName(), preferenceList);
     }
     LOG.info(String
         .format("End computing ideal state for resource %s in maintenance mode.", resourceName));
     return currentIdealState;
   }
+
+  public static void main(String[] args) {
```

Comment:
any reason to have 'main' in this file?

Reply:
Sorry, this is not intended.. Removed now!